darshan-util/pydarshan/darshan/cli/summary.py

```diff
@@ -0,0 +1,645 @@
+    df_min_rank = df_apmpi.loc[df_apmpi["Rank"] == min_rank]
+    df_mean_rank = df_apmpi.loc[df_apmpi["Rank"] == mean_rank]
+    df_zero_rank = df_apmpi.loc[df_apmpi["Rank"] == 0]
+    #encoded.append(df_max_rank.round({'Total_Time':6, 'Count':6, 'Total_Bytes':6, '[0-256B]':6, '[256-1KB]':6, '[1K-8KB]':6, '[8K-256KB]':6, '[256K-1MB]':6, '[1MB+]':6, 'Min_Time':6, 'Max_Time':6}))
+    #encoded.append(df_min_rank.round({'Total_Time':6, 'Count':6, 'Total_Bytes':6, '[0-256B]':6, '[256-1KB]':6, '[1K-8KB]':6, '[8K-256KB]':6, '[256K-1MB]':6, '[1MB+]':6, 'Min_Time':6, 'Max_Time':6}))
+    #encoded.append(df_mean_rank.round({'Total_Time':6, 'Count':6, 'Total_Bytes':6, '[0-256B]':6, '[256-1KB]':6, '[1K-8KB]':6, '[8K-256KB]':6, '[256K-1MB]':6, '[1MB+]':6, 'Min_Time':6, 'Max_Time':6}))
+    #encoded.append(df_zero_rank.round({'Total_Time':6, 'Count':6, 'Total_Bytes':6, '[0-256B]':6, '[256-1KB]':6, '[1K-8KB]':6, '[8K-256KB]':6, '[256K-1MB]':6, '[1MB+]':6, 'Min_Time':6, 'Max_Time':6}))
+    encoded.append(df_max_rank)
+    encoded.append(df_min_rank)
+    encoded.append(df_mean_rank)
+    encoded.append(df_zero_rank)
+    print(f'Phase7: {time.process_time() - start}')
+    return encoded
+  
+def main(args=None):
+
+    if args is None:
+        parser = argparse.ArgumentParser(description='')
+        setup_parser(parser)
+        args = parser.parse_args()
+
+    if args.debug:
+        print(args)
+    
+    init_plotting()
+    variables = {}
+    report = darshan.DarshanReport(args.input, read_all=True)
+    #report.info()
+
+    #
+    # Setup template header variabels
+    #
+    variables['exe'] = report.metadata['exe']
+    variables['date'] = datetime.datetime.fromtimestamp(report.metadata['job']['start_time'], pytz.utc)
+    variables['jid'] = report.metadata['job']['jobid']
+    variables['uid'] = report.metadata['job']['uid']
+    variables['nprocs'] = report.metadata['job']['nprocs']
+    etime = int(report.metadata['job']['end_time'])
+    stime = int(report.metadata['job']['start_time'])
+    if etime > stime:
+        variables['runtime'] = etime - stime + 1
+    else:
+        variables['runtime'] = 0
+
+    if 'POSIX' in report.modules:
+        posix_df = report.records['POSIX'].to_df()
+    else:
+        posix_df = None
+
+    if 'MPI-IO' in report.modules:
+        mpiio_df = report.records['MPI-IO'].to_df()
+    else:
+        mpiio_df = None
+    
+    if 'STDIO' in report.modules:
+        stdio_df = report.records['STDIO'].to_df()
+    else:
+        stdio_df = None
+    
+    if 'APXC' in report.modules:
+        apxc_dict = report.records['APXC'].to_dict()
+    else:
+        apxc_dict = None
+
+    if 'APMPI' in report.modules:
+        apmpi_dict = report.records['APMPI'].to_dict()
+    else:
+        apmpi_dict = None
+
+    #
+    # Plot I/O cost
+    #
+    variables['plot_io_cost'] = plot_io_cost(posix_df, mpiio_df, stdio_df, int(variables['runtime']), int(variables['nprocs'])).decode('utf-8')
```

Comment:
It is somewhat debatable how flexible this integration system is--for example of hacking in another plot into the report see: https://github.com/darshan-hpc/darshan/pull/396#issuecomment-866323498

It probably wouldn't be all that bad if you could jettison the html stuff that surrounds the framework.

Reply:
Or, programmatically inject the appropriate html for a new tracing module when it is leveraged.